core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -877,6 +877,10 @@ public RelSubset ensureRegistered(RelNode rel, RelNode equivRel) {
    * Checks internal consistency.
    */
   protected boolean isValid(Litmus litmus) {
+    if (this.getRoot() == null) {
+      return true;
```

Comment:
-1 for this fix

Reply:
BTW, next time we fix the code, can we try to give some tests ? Without the test, no one can make sure how the code works and why it should be fixed like that.

Reply:
Thanks for reviewing. We can revert the change and reopen the issue if you would like to.

Reply:
The test step is documented in JIRA. You turn on DEBUG and then you can verify. We had a discussion about using DEBUG as default option in build test. But this idea was turned down due to other concerns (also mentioned in JIRA). 

In this specific case, what's your reason of -1?

Reply:
I have left the comment in the commit page.

Reply:
I'm not very familiar with VolcanoPlanner implementation details, but I have the same impression as @danny0405 ; I think that in the case of no root, we could get the cluster from somewhere else (maybe from the first subset that we process?).

Reply:
This if branch is only executed when isValid() is called during VolcanoPlanner.setRoot(). At the moment, although RelNodes and RelSubsets are created by the registerImpl(), there's no point of doing validation since this is at the very beginning volcano planning, all of the nodes are logical with infinite cost, and the memo is plain empty.

Reply:
I don't think so, at least these cases would invoke the `isValid`:
* When user change the node traits manually [1]
* When we generate a new node in the planner rule and register it into the planner [2]
* Some explicit invocation of the node registration

[1] https://github.com/apache/calcite/blob/301bbe9e132346f60587e8c21c474f7c51b236c3/core/src/main/java/org/apache/calcite/plan/RelOptRule.java#L568

[2] https://github.com/apache/calcite/blob/301bbe9e132346f60587e8c21c474f7c51b236c3/core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java#L132

[3] https://github.com/apache/calcite/blob/301bbe9e132346f60587e8c21c474f7c51b236c3/core/src/main/java/org/apache/calcite/rel/rules/JoinCommuteRule.java#L176

Reply:
BTW, this code `this.getRoot().getCluster().getMetadataQuerySupplier().get()` confused me a lot, can you please explain why we need a fresh new `RelMetadataQuery` instance here ? Couldn't we use the instance already existing there with `RelOptCluster#getMetadataQuery` ? (I have checked that almost each `isValid` case is in the `RelOptRuleCall` circle).

Even we have to got a fresh new instance here, why we just add a new interface `RelOptCluster#getMetadataQuerySupplier` just for debugging ? Can you refactor that out ?

Reply:
All of the cases you mentioned are not related to the scenario where root is null, so they are irrelevant to this patch. 

Regarding why we need a new instance, we want to make sure the validation does not incur any side effects, so a new instance of meta data query won't mess with the existing meta data cache. This was discussed before in the JIRA, 

Regarding getMetadataQuerySupplier, please see CALCITE-3487.

Reply:
Then can you clarify you scenario where the root is null, there is already code there we can fetch the cluster instance and why in your code just skip the null root case check ?

I'm still agains to add a top level interface just for DEBUGGING.

Reply:
I am not sure I understand your question. I didn't still the check. The patch adds the check.

The interface is not for debugging. See CALCITE-3487.